include/onnxruntime/core/common/span_utils.h

```diff
@@ -64,7 +67,22 @@ constexpr auto AsSpan(const T (&arr)[N]) {
   return details::AsSpanImpl(arr, N);
 }
 
-template<class T>
+template <class T>
 inline gsl::span<const T> EmptySpan() { return gsl::span<const T>(); }
 
-}
\ No newline at end of file
+template <class U, class T>
+[[nodiscard]] inline gsl::span<U> ReinterpretAsSpan(gsl::span<T> src) {
+  // adapted from gsl-lite span::as_span():
+  // https://github.com/gsl-lite/gsl-lite/blob/4720a2980a30da085b4ddb4a0ea2a71af7351a48/include/gsl/gsl-lite.hpp#L4102-L4108
+  Expects(src.size_bytes() % sizeof(U) == 0);
+  return gsl::span<U>(reinterpret_cast<U*>(src.data()), src.size_bytes() / sizeof(U));
+}
+
+template <class T1, size_t Extent1, class T2, size_t Extent2>
+[[nodiscard]] inline bool SpanEq(gsl::span<T1, Extent1> a, gsl::span<T2, Extent2> b) {
```

Comment:
gsl::span comparison operator differences vs. gsl-lite:
- does not support comparing `span<T>` with `span<const T>`
- span comparison operators are disabled entirely in an ORT build without exceptions
  - https://github.com/microsoft/GSL/blob/10df83d292bf5bbdc487e57dc8c2dc8c7a01f4d1/include/gsl/span_ext#L49-L51
  - https://github.com/microsoft/GSL/blob/10df83d292bf5bbdc487e57dc8c2dc8c7a01f4d1/include/gsl/assert#L24-L25

hence this helper to replace operator==().

note that operator==() enables nicer GTEST equality assertion messages, in addition to more natural looking code like `span1 == span2`.

note that std::span (C++20) also does not provide comparison operators.